File: examples/dot_product.c
/* dot_product.c — Dot product (reduction loop)
 *
 * Parallelism type: POLYHEDRAL
 * Reduction variable: sum (safe with OpenMP reduction clause)
 * Expected transform: #pragma omp parallel for reduction(+:sum)
 */
float dot_product(float* A, float* B, int N) {
    float sum = 0.0f;
    for (int i = 0; i < N; i++) {
        sum += A[i] * B[i];
    }
    return sum;
}

File: examples/rust/dot_product.rs
// APG Example: dot_product (Rust/Rayon)
// Computes the dot product of two slices A and B.
// Expected parallelism: reduction (parallel fold + sum)
pub fn dot_product(a: &[f32], b: &[f32], n: usize) -> f32 {
    let mut s: f32 = 0.0;
    for i in 0..n {
        s += a[i] * b[i];
    }
    s
}
